File: IwashiScopePackage/Sources/IwashiScopeFeature/Models/ColorAppearance/SpectralTristimulusIntegrator.swift
import Foundation

/// Shared, relative-Y reflectance integration; never uses a chart's clipped display range.
enum SpectralTristimulusIntegrator {
    struct IntegrationResult {
        let objectXYZ: Vector3
        let whiteXYZ: Vector3
        let wavelengths: [Double]
    }

    static func integrate(
        reflectance: [SpectralSample],
        illuminant: [SpectralSample],
        wavelengths: [Double]? = nil
    ) -> IntegrationResult? {
        let orderedReflectance = reflectance
            .filter {
                $0.wavelength.isFinite && $0.value.isFinite
            }
            .sorted { $0.wavelength < $1.wavelength }
        let orderedIlluminant = illuminant
            .filter {
                $0.wavelength.isFinite
                    && $0.value.isFinite
                    && $0.value >= 0
            }
            .sorted { $0.wavelength < $1.wavelength }
        guard orderedReflectance.count >= 2,
              orderedIlluminant.count >= 2,
              zip(orderedReflectance, orderedReflectance.dropFirst()).allSatisfy({ $0.wavelength < $1.wavelength }),
              zip(orderedIlluminant, orderedIlluminant.dropFirst()).allSatisfy({ $0.wavelength < $1.wavelength }) else {
            return nil
        }

        var whiteX = 0.0
        var whiteY = 0.0
        var whiteZ = 0.0
        var objectX = 0.0
        var objectY = 0.0
        var objectZ = 0.0
        var usedWavelengths: [Double] = []

        for wavelength in wavelengths ?? Array(stride(from: 380.0, through: 730.0, by: 5.0)) {
            guard let source = interpolatedValue(
                      in: orderedIlluminant,
                      at: wavelength
                  ),
                  let observer = observerValues(at: wavelength),
                  let reflectanceValue = interpolatedValue(
                      in: orderedReflectance,
                      at: wavelength
                  ) else {
                continue
            }

            let scaledReflectance = max(0, reflectanceValue) / 100
            whiteX += source * observer.x
            whiteY += source * observer.y
            whiteZ += source * observer.z
            objectX += source * scaledReflectance * observer.x
            objectY += source * scaledReflectance * observer.y
            objectZ += source * scaledReflectance * observer.z
            usedWavelengths.append(wavelength)
        }

        guard usedWavelengths.count >= 2,
              whiteY.isFinite,
              whiteY > 1e-12 else {
            return nil
        }

        let normalization = 100 / whiteY
        let whiteXYZ = Vector3(
            first: whiteX * normalization,
            second: 100,
            third: whiteZ * normalization
        )
        let objectXYZ = Vector3(
            first: objectX * normalization,
            second: objectY * normalization,
            third: objectZ * normalization
        )
        guard [
            whiteXYZ.first, whiteXYZ.second, whiteXYZ.third,
            objectXYZ.first, objectXYZ.second, objectXYZ.third,
        ].allSatisfy(\.isFinite) else {
            return nil
        }
        return IntegrationResult(objectXYZ: objectXYZ, whiteXYZ: whiteXYZ, wavelengths: usedWavelengths)
    }

    private static func observerValues(
        at wavelength: Double
    ) -> (x: Double, y: Double, z: Double)? {
        let indexValue = (
            wavelength - ColorRenderingReferenceData.startWavelength
        ) / ColorRenderingReferenceData.interval
        let index = Int(indexValue.rounded())
        guard abs(indexValue - Double(index)) < 1e-9,
              ColorRenderingReferenceData.xBar.indices.contains(index),
              ColorRenderingReferenceData.yBar.indices.contains(index),
              ColorRenderingReferenceData.zBar.indices.contains(index) else {
            return nil
        }
        return (
            ColorRenderingReferenceData.xBar[index],
            ColorRenderingReferenceData.yBar[index],
            ColorRenderingReferenceData.zBar[index]
        )
    }

    private static func interpolatedValue(
        in samples: [SpectralSample],
        at wavelength: Double
    ) -> Double? {
        guard let first = samples.first,
              let last = samples.last,
              wavelength >= first.wavelength,
              wavelength <= last.wavelength else {
            return nil
        }
        if wavelength == first.wavelength { return first.value }
        if wavelength == last.wavelength { return last.value }

        var lowerIndex = 0
        var upperIndex = samples.count - 1
        while lowerIndex + 1 < upperIndex {
            let midpoint = (lowerIndex + upperIndex) / 2
            if samples[midpoint].wavelength <= wavelength {
                lowerIndex = midpoint
            } else {
                upperIndex = midpoint
            }
        }

        let lower = samples[lowerIndex]
        let upper = samples[upperIndex]
        let width = upper.wavelength - lower.wavelength
        guard width > 0 else { return nil }
        let fraction = (wavelength - lower.wavelength) / width
        return lower.value + ((upper.value - lower.value) * fraction)
    }
}

File: Windows/src/IwashiScope.Core/Calculations/SpectralTristimulusIntegrator.cs
using IwashiScope.Core.Models;

namespace IwashiScope.Core.Calculations;

public static class SpectralTristimulusIntegrator
{
    public sealed record IntegrationResult(Vector3 ObjectXyz, Vector3 WhiteXyz, IReadOnlyList<double> Wavelengths);
    private static readonly double[] DefaultWavelengths = Enumerable.Range(0, 71).Select(i => 380d + i * 5).ToArray();

    public static IntegrationResult? Integrate(
        IReadOnlyList<SpectralSample> reflectance,
        IReadOnlyList<SpectralSample> illuminant,
        IReadOnlyList<double>? wavelengths = null)
    {
        var orderedReflectance = reflectance
            .Where(s => double.IsFinite(s.Wavelength) && double.IsFinite(s.Value))
            .OrderBy(s => s.Wavelength).ToArray();
        var orderedIlluminant = illuminant
            .Where(s => double.IsFinite(s.Wavelength) && double.IsFinite(s.Value) && s.Value >= 0)
            .OrderBy(s => s.Wavelength).ToArray();
        if (orderedReflectance.Length < 2 || orderedIlluminant.Length < 2 ||
            orderedReflectance.Zip(orderedReflectance.Skip(1)).Any(p => p.First.Wavelength >= p.Second.Wavelength) ||
            orderedIlluminant.Zip(orderedIlluminant.Skip(1)).Any(p => p.First.Wavelength >= p.Second.Wavelength))
            return null;

        double whiteX = 0, whiteY = 0, whiteZ = 0, objectX = 0, objectY = 0, objectZ = 0;
        var used = new List<double>();
        foreach (var wavelength in wavelengths ?? DefaultWavelengths)
        {
            var source = ReflectanceIlluminantSpectrumCalculator.Interpolate(orderedIlluminant, wavelength);
            var reflected = ReflectanceIlluminantSpectrumCalculator.Interpolate(orderedReflectance, wavelength);
            var observer = ObserverValues(wavelength);
            if (source is null || reflected is null || observer is null) continue;
            var scaledReflectance = Math.Max(0, reflected.Value) / 100;
            whiteX += source.Value * observer.Value.X;
            whiteY += source.Value * observer.Value.Y;
            whiteZ += source.Value * observer.Value.Z;
            objectX += source.Value * scaledReflectance * observer.Value.X;
            objectY += source.Value * scaledReflectance * observer.Value.Y;
            objectZ += source.Value * scaledReflectance * observer.Value.Z;
            used.Add(wavelength);
        }
        if (used.Count < 2 || !double.IsFinite(whiteY) || whiteY <= 1e-12) return null;
        var normalization = 100 / whiteY;
        var white = new Vector3(whiteX * normalization, 100, whiteZ * normalization);
        var value = new Vector3(objectX * normalization, objectY * normalization, objectZ * normalization);
        return white.IsFinite && value.IsFinite ? new IntegrationResult(value, white, used.ToArray()) : null;
    }

    private static (double X, double Y, double Z)? ObserverValues(double wavelength)
    {
        var indexValue = (wavelength - ColorRenderingReferenceData.StartWavelength) / ColorRenderingReferenceData.Interval;
        if (!double.IsFinite(indexValue)) return null;
        var index = (int)Math.Round(indexValue);
        if (Math.Abs(indexValue - index) >= 1e-9 || index < 0 ||
            index >= ColorRenderingReferenceData.XBar.Length ||
            index >= ColorRenderingReferenceData.YBar.Length ||
            index >= ColorRenderingReferenceData.ZBar.Length) return null;
        return (ColorRenderingReferenceData.XBar[index], ColorRenderingReferenceData.YBar[index], ColorRenderingReferenceData.ZBar[index]);
    }
}
